packages/web3-eth/src/utils/get_transaction_error.ts

```diff
@@ -0,0 +1,84 @@
+/*
+This file is part of web3.js.
+
+web3.js is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+web3.js is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { Web3Context } from 'web3-core';
+import {
+	TransactionRevertedWithoutReasonError,
+	TransactionRevertInstructionError,
+	TransactionRevertWithCustomError,
+} from 'web3-errors';
+import { ContractAbi, TransactionCall, TransactionReceipt } from 'web3-types';
+import { DataFormat, FormatType } from 'web3-utils';
+import { RevertReason, RevertReasonWithCustomError } from '../types';
+// eslint-disable-next-line import/no-cycle
+import { getRevertReason, parseTransactionError } from './get_revert_reason';
+
+export async function getTransactionError<ReturnFormat extends DataFormat>(
+	web3Context: Web3Context,
+	transactionFormatted?: TransactionCall,
+	transactionReceiptFormatted?: FormatType<TransactionReceipt, ReturnFormat>,
+	receivedError?: unknown,
+	contractAbi?: ContractAbi,
+) {
+	let reason: string | RevertReason | RevertReasonWithCustomError | undefined;
+
+	if (receivedError !== undefined) {
+		reason = parseTransactionError(receivedError);
+	} else if (web3Context.handleRevert && transactionFormatted !== undefined) {
```

Comment:
As shared above, eth_call with tx obj should be sent before actual tx and this function is used after sending tx.